packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -139,112 +201,139 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
       case 'Enter':
       case ' ':
       case 'Spacebar': {
-        !disabled && onClick?.(value);
+        onClick();
         event.preventDefault();
         break;
       }
-      case 'ArrowRight': {
-        if (subMenuItems.length > 0) {
-          setIsSubmenuVisible(true);
+      default:
+        break;
+    }
+  };
 
-          // flush and reset state so we are ready to focus again next time
-          flushSync(() => setFocusOnSubmenu(true));
-          setFocusOnSubmenu(false);
+  const onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
+    // Focus the item when hovered.
+    if (e.target === e.currentTarget) {
+      menuItemRef.current?.focus();
 
-          event.preventDefault();
-          event.stopPropagation();
-        }
-        break;
+      // Since we manually focus the MenuItem on hover, we need to manually update the active index for
+      // Floating UI's keyboard navigation to work correctly.
+      if (parent != null && indexInParentTree != null) {
+        parent.setActiveIndex(indexInParentTree);
       }
-      case 'ArrowLeft': {
-        if (parent.ref) {
-          parent.ref.current?.focus();
-          parent.setIsNestedSubmenuVisible(false);
-        }
-        event.stopPropagation();
-        event.preventDefault();
-        break;
-      }
-      case 'Escape': {
-        // focus might get lost if submenu closes so move it back to parent
-        parent.ref?.current?.focus();
-        break;
-      }
-      default:
-        break;
     }
```

Comment:
@mayank99 and I had removed this in a previous commit (https://github.com/iTwin/iTwinUI/pull/1942/commits/3861266a09f6d06d6a5f79e6e77261b56b12f309). But I brought it back since I now remember why I had it in the first place. It was because if we don't update the active index upon hover/focus, then Floating UI's `useListNavigation` keeps the stale `activeIndex`.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/29fb5fc2-3ce0-4761-991f-6800f6edca9b

</details>

One small difference is that instead of updating the parent's activeIndex in `onFocus`, I am now doing it in `onMouseEnter`, where it is really needed because that is where we're manually changing the focus instead of letting `useListNavigation` handle it.